File: Trivia/game.c
#include <stdio.h>
#include <string.h>
#include <malloc.h>
#include "game.h"

void game_initialize(game_t *self)
{
	int i;
	memset(self, 0, sizeof(game_t));

	for (i = 0; i < 50; ++i)
	{
		sprintf(self->pop_questions[i], "Pop Question %d", i);
		sprintf(self->science_questions[i], "Science Question %d", i);
		sprintf(self->sports_questions[i], "Sports Question %d", i);

		char *rock_question = game_create_rock_question(i);
		strcpy(self->rock_questions[i], rock_question);
		free(rock_question);
	}
}

char* game_create_rock_question(int index)
{
	char index_str[127];
	sprintf(index_str, "Rock Question %d", index);
	return strdup(index_str);
}

bool game_is_playable(game_t *self)
{
	return (game_how_many_players(self) >= 2);
}

void game_ask_question(game_t *self)
{
	if (strcmp(game_current_category(self), "Pop") == 0)
	{
		printf("%s\n", self->pop_questions[self->cur_pop++]);

	}
	if (strcmp(game_current_category(self), "Science") == 0)
	{
		printf("%s\n", self->science_questions[self->cur_sci++]);

	}
	if (strcmp(game_current_category(self), "Sports") == 0)
	{
		printf("%s\n", self->sports_questions[self->cur_spo++]);

	}
	if (strcmp(game_current_category(self), "Rock") == 0)
	{
		printf("%s\n", self->rock_questions[self->cur_roc++]);

	}
}

bool game_add(game_t *self, const char *name)
{
	self->players[self->cur++] = strdup(name);
	self->places[self->cur] = 0;
	self->purses[self->cur] = 0;
	self->in_penalty_box[self->cur] = false;

	printf("%s was added\n", name);
	printf("They are player number %d\n", self->cur);

	return true;
}

void game_roll(game_t *self, int die)
{

}

int game_how_many_players(game_t *self)
{
	return self->cur;
}

bool game_was_correctly_answered(game_t *self)
{
	if (self->in_penalty_box[self->current_player])
	{
		if (self->out_of_penalty_box)
		{
			printf("Answer was correct!!!!\n");
			self->purses[self->current_player]++;
			printf("%s now has %d Gold Coins.\n",
					self->players[self->current_player],
					self->purses[self->current_player]);

			bool winner = game_did_player_win(self);
			self->current_player++;
			if (self->current_player == self->cur)
			{
				self->current_player = 0;
			}

			return winner;
		}
		else
		{
			self->current_player++;
			if (self->current_player == self->cur)
			{
				self->current_player = 0;
			}

			return true;
		}
	}
	else
	{
		printf("Answer was correct!!!!\n");
		self->purses[self->current_player]++;
		printf("%s now has %d Gold Coins.\n",
				self->players[self->current_player],
				self->purses[self->current_player]);

		bool winner = game_did_player_win(self);
		self->current_player++;
		if (self->current_player == self->cur)
		{
			self->current_player = 0;
		}

		return winner;
	}
}

bool game_wrong_answer(game_t *self)
{
	printf("Question was incorrectly answered\n");
	printf("%s was sent to the penalty box\n",
			self->players[self->current_player]);
	self->in_penalty_box[self->current_player] = true;

	self->current_player++;
	if (self->current_player == self->cur)
	{
		self->current_player = 0;
	}
	return true;
}

bool game_did_player_win(game_t *self)
{
	return !(self->purses[self->current_player] == 6);
}

const char * game_current_category(game_t *self)
{
	if (self->places[self->current_player] == 0)
		return "Pop";
	if (self->places[self->current_player] == 4)
		return "Pop";
	if (self->places[self->current_player] == 8)
		return "Pop";
	if (self->places[self->current_player] == 1)
		return "Science";
	if (self->places[self->current_player] == 5)
		return "Science";
	if (self->places[self->current_player] == 9)
		return "Science";
	if (self->places[self->current_player] == 2)
		return "Sports";
	if (self->places[self->current_player] == 6)
		return "Sports";
	if (self->places[self->current_player] == 10)
		return "Sports";
	return "Rock";
}

File: C#/Trivia/Trivia/Game.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace UglyTrivia
{
    public class Game
    {
        List<string> players = new List<string>();

        int[] playerLocations = new int[6];
        int[] purses = new int[6];

        bool[] inPenaltyBox = new bool[6];

        LinkedList<string> popQuestions = new LinkedList<string>();
        LinkedList<string> scienceQuestions = new LinkedList<string>();
        LinkedList<string> sportsQuestions = new LinkedList<string>();
        LinkedList<string> rockQuestions = new LinkedList<string>();

        int currentPlayerIndex = 0;
        bool isGettingOutOfPenaltyBox;

        public Game()
        {
            for (int i = 0; i < 50; i++)
            {
                popQuestions.AddLast("Pop Question " + i);
                scienceQuestions.AddLast("Science Question " + i);
                sportsQuestions.AddLast("Sports Question " + i);
                rockQuestions.AddLast(createRockQuestion(i));
            }
        }

        public String createRockQuestion(int index)
        {
            return "Rock Question " + index;
        }

        public bool HasEnoughPlayers()
        {
            return (howManyPlayers() >= 2);
        }

        public bool add(String playerName)
        {
            players.Add(playerName);
            var addedPlayerIndex = howManyPlayers();
            playerLocations[addedPlayerIndex] = 0;
            purses[addedPlayerIndex] = 0;
            inPenaltyBox[addedPlayerIndex] = false;

            Console.WriteLine(playerName + " was added");
            Console.WriteLine("They are player number " + players.Count);
            return true;
        }

        public int howManyPlayers()
        {
            return players.Count;
        }

        public void MovePlayerAndAskQuestion(int rolledNumber)
        {
            Console.WriteLine(players[currentPlayerIndex] + " is the current player");
            Console.WriteLine("They have rolled a " + rolledNumber);

            if (inPenaltyBox[currentPlayerIndex])
            {
                if (rolledNumber % 2 != 0)
                {
                    isGettingOutOfPenaltyBox = true;

                    Console.WriteLine(players[currentPlayerIndex] + " is getting out of the penalty box");
                    playerLocations[currentPlayerIndex] = playerLocations[currentPlayerIndex] + rolledNumber;
                    if (playerLocations[currentPlayerIndex] > 11) playerLocations[currentPlayerIndex] = playerLocations[currentPlayerIndex] - 12;

                    Console.WriteLine(players[currentPlayerIndex]
                            + "'s new location is "
                            + playerLocations[currentPlayerIndex]);
                    Console.WriteLine("The category is " + currentCategory());
                    askQuestion();
                }
                else
                {
                    Console.WriteLine(players[currentPlayerIndex] + " is not getting out of the penalty box");
                    isGettingOutOfPenaltyBox = false;
                }
            }
            else
            {
                playerLocations[currentPlayerIndex] = playerLocations[currentPlayerIndex] + rolledNumber;
                if (playerLocations[currentPlayerIndex] > 11) playerLocations[currentPlayerIndex] = playerLocations[currentPlayerIndex] - 12;

                Console.WriteLine(players[currentPlayerIndex]
                        + "'s new location is "
                        + playerLocations[currentPlayerIndex]);
                Console.WriteLine("The category is " + currentCategory());
                askQuestion();
            }
        }

        private void askQuestion()
        {
            if (currentCategory() == "Pop")
            {
                Console.WriteLine(popQuestions.First());
                popQuestions.RemoveFirst();
            }
            if (currentCategory() == "Science")
            {
                Console.WriteLine(scienceQuestions.First());
                scienceQuestions.RemoveFirst();
            }
            if (currentCategory() == "Sports")
            {
                Console.WriteLine(sportsQuestions.First());
                sportsQuestions.RemoveFirst();
            }
            if (currentCategory() == "Rock")
            {
                Console.WriteLine(rockQuestions.First());
                rockQuestions.RemoveFirst();
            }
        }

        private String currentCategory()
        {
            if (playerLocations[currentPlayerIndex] == 0) return "Pop";
            if (playerLocations[currentPlayerIndex] == 4) return "Pop";
            if (playerLocations[currentPlayerIndex] == 8) return "Pop";
            if (playerLocations[currentPlayerIndex] == 1) return "Science";
            if (playerLocations[currentPlayerIndex] == 5) return "Science";
            if (playerLocations[currentPlayerIndex] == 9) return "Science";
            if (playerLocations[currentPlayerIndex] == 2) return "Sports";
            if (playerLocations[currentPlayerIndex] == 6) return "Sports";
            if (playerLocations[currentPlayerIndex] == 10) return "Sports";
            return "Rock";
        }

        public bool wasCorrectlyAnswered()
        {
            if (inPenaltyBox[currentPlayerIndex])
            {
                if (isGettingOutOfPenaltyBox)
                {
                    Console.WriteLine("Answer was correct!!!!");
                    purses[currentPlayerIndex]++;
                    Console.WriteLine(players[currentPlayerIndex]
                            + " now has "
                            + purses[currentPlayerIndex]
                            + " Gold Coins.");

                    bool winner = didPlayerWin();
                    currentPlayerIndex++;
                    if (currentPlayerIndex == players.Count) currentPlayerIndex = 0;

                    return winner;
                }
                else
                {
                    currentPlayerIndex++;
                    if (currentPlayerIndex == players.Count) currentPlayerIndex = 0;
                    return true;
                }
            }
            else
            {
                Console.WriteLine("Answer was corrent!!!!");
                purses[currentPlayerIndex]++;
                Console.WriteLine(players[currentPlayerIndex]
                        + " now has "
                        + purses[currentPlayerIndex]
                        + " Gold Coins.");

                bool winner = didPlayerWin();
                currentPlayerIndex++;
                if (currentPlayerIndex == players.Count) currentPlayerIndex = 0;

                return winner;
            }
        }

        public bool wrongAnswer()
        {
            Console.WriteLine("Question was incorrectly answered");
            Console.WriteLine(players[currentPlayerIndex] + " was sent to the penalty box");
            inPenaltyBox[currentPlayerIndex] = true;

            currentPlayerIndex++;
            if (currentPlayerIndex == players.Count) currentPlayerIndex = 0;
            return true;
        }

        private bool didPlayerWin()
        {
            return !(purses[currentPlayerIndex] == 6);
        }
    }
}
